### ISSUE TITLE
Data in a box/container

### ISSUE THREAD

**Issue by JPXKQX** [ecmwf/anemoi-core#738]
### Is your feature request related to a problem? Please describe.

Currently, models in `anemoi-core` expect raw `torch.Tensor` objects in their `forward()` method. This design works well with the current functionality but could limit flexibility when handling more complex use cases (e.g. multi-domain training, ...). I raise this for discussion about the different options/designs to refactor the model API to use a dict-like structured container (potentially tensordict, a lightweight custom dict, a regular dict, or other) that encapsulates both the tensor payloads and their associated metadata. 

### Motivation 

As the framework evolves toward more complex training workflows, particularly multi-domain training, alternating between datasets, and integrating heterogeneous variable sets, the current API becomes a bottleneck: 

1. **Lack of metadata propagation**
     Many modelling tasks require metadata such as: 
     - variable names and ordering 
     - dataset- or domain-specific statistics 
     - normalization parameters (statistics -> processors) 
     - coordinate information (latlons -> graph) 

     Currently, the alternative would be to build dicts during model initialization, which leads to brittle code paths and breaks separation of concerns. 

2. **Limited extensibility** 
As new tasks require richer contextual information, simply passing raw tensors forces the model to infer too much from too little information. This could play an important role in providing capabilities for masked training or more complex encoder/decoder configurations in the future, using not only the raw values, but also pressure level information through positional encoding, for example. 

### Describe the solution you'd like

Refactor the core model interface so that: 

`forward()` accepts a dict-like object, (a box/container) e.g.: 
```
{ 
    “data”: <torch.Tensor>, 

    “latitudes”: <torch.Tensor>, 
    “name_to_index”: [...], 
    “statistics”: {...}, 
    ... 
} 
```
- Specifically, this should be a dict of dicts after multiple datasets is merged. 

- Optionally, use tensordict (https://docs.pytorch.org/tensordict/stable/index.html) for better nested structure support, device movement, and JIT compatibility, but it’s still 0.X version. 
 Adjust dataloaders so they natively produce these structured objects before entering the model, instead of assembling metadata inside the model initialization. 

### Benefits 

- Enables multi-domain training where each sample or batch carries its own metadata without modifying the model. 
- Improves clarity and maintainability: the model receives a fully-described sample rather than relying on global or implicit assumptions. 
- Supports future heterogeneous modeling tasks, including: 
      - multiple variable subsets 
- Makes models more reusable and composable across multiple data sources and tasks. 
- Enabling dynamic imputing by carrying the masks in the box/container. 

### Implementation  

Different options for the implementation of the data container: 

- **Python dict**. It’s simple, flexible, and familiar, but lacks built-in mechanisms for device movement and batch-level operations. 

- **A custom structured dict-like class** (e.g., AnemoiBatch, AnemoiDict, AnemoiTensor, AnemoiTensorDict ...). This could enforce invariants, provide typed accessors, and help avoid inconsistent key naming, at the cost of additional maintenance. 

- **TensorDict**. It offers native support for nested tensor structures, batched operations, device transfers, and TorchScript-friendliness, but it’s still under development (version 0.8). 

### Transition Plan 

- Introduce the new interface (box/container) by moving the actual torch.Tensor to {“data”: torch.Tensor}. What is the best key? data? values? Tensor? 
- Introduce different information into the box 
    - Introduce latitudes and longitudes. They should be the same. They should be used by the GraphProvider’s in the model 
    - Introduce statistics & statistics_tendencies 
    - Introduce the dimensions: -> {...., “dims”: (“batch”, “ensemble”, “grid”, “time”, “variables”)}. "time" is ok, but it is actually closer to "time_offsets". "grid" is too specific. "latlons" also (thinking about spherical harmonics), "values" may be a good compromise. 
    - ... 

### Questions

- How deep should the box/container go?
- What functionality do we want to re-use in `anemoi-inference`?
- What keys/information should the box contain?

**Comment by anaprietonem** [comment]
https://github.com/ecmwf/anemoi-core/pull/831

**Comment by frazane** [comment]
Just sharing here that TensorDict is quite actively developed (whether it's a good or bad thing is up to you), see the [latest release](https://github.com/pytorch/tensordict/releases/tag/v0.13.0) and the impressive pace of the [PRs queue](https://github.com/pytorch/tensordict/pulls).

**Comment by JPXKQX** [comment]
Update on current status:

The class `Batch` is a [frozen dataclass](https://github.com/ecmwf/anemoi-core/blob/feat/richer-batch/models/src/anemoi/models/data/batch.py#L125) whose job is to hold multiple datasets (gridded and sparse observations) side by side and expose them through one consistent interface. It is a dict-of-datasets. Every field is keyed by dataset name, so one `Batch` can hold, say, "era5" (gridded) and "obs" (sparse observations) at once, each with its own data, shape, layout and ... (we call this bundle of objects a `SourceView`).

**Comment by mchantry** [comment]
Thanks @JPXKQX what's the current view on this question?
> What functionality do we want to re-use in anemoi-inference?
